Improve screenshot path handling

veracioux opened the pull request:
This is a work in progress. Tracks issue #1784

**Some notes:**
- The `format` argument to `properScreenshotPath` was added to predict for issue #1785.
- The reason I decided to rework `FileNameHandler` is because I wanted something easier to work with. And the previous implementation was confusing.

mmahmoudian commented on src/utils/filenamehandler.cpp:
```diff
@@ -52,70 +52,71 @@ QString FileNameHandler::parseFilename(const QString& name)
-          QStandardPaths::writableLocation(QStandardPaths::PicturesLocation);
+    QFileInfo info(path);
+    if (info.isDir()) {
+        return QDir(QDir(path).absolutePath() + "/" + parsedPattern() + "." +
```
Is using the "/" in the file path valid for Windows?

veracioux replied:
QDir should convert the delimiter automatically.

mmahmoudian commented:
![image](https://user-images.githubusercontent.com/390889/128860198-73af396a-e700-4df3-8c0c-296911ce2715.png)

mmahmoudian commented:
I tried with the AppImage from this PR, and also commented out all lines of `path_option.sh` that were not necessary in my opinion:

<details>
<summary>My version of <code>path_option.sh</code></summary>

```sh
#!/usr/bin/env sh

# pkill -i flameshot
# flameshot &
# 
# sleep 4
# 
# # Create a directory for file output
# rm -rf _out 2>/dev/null
# mkdir -p _out
# cd _out
# 
# # NOTE THIS
# # The dir _out will be have as the home directory
# HOME="$PWD"
# mkdir -p subdir
# 
echo ">> Nonexistent directory is not a valid path"
flameshot screen -p ~/blah/blah

sleep 2
echo ">> Saved relative to PWD"
flameshot screen -p relative.png

sleep 2
echo ">> Absolute paths work too"
flameshot screen -p ~/absolute.png

sleep 2
echo ">> Redundancy in the path will be removed"
flameshot screen -p ~/subdir/..///redundancy_removed.png

sleep 2
echo ">> If the destionation is a directory, the file name is generated from strf from the config"
flameshot screen -p ./

sleep 2
echo ">> If the output file has no suffix, it will be added"
flameshot screen -p ~/without_suffix

sleep 2
echo ">> If the suffix does not match the output format, it will be overwritten"
flameshot screen -p ~/wrong_suffix.jpg

sleep 2
echo ">> If the destination path exists, it will have _NUM appended to the base name"
flameshot screen -p ~/absolute.png

sleep 2
echo ">> We will do that again but without a suffix"
flameshot screen -p ~/absolute

sleep 2

# sleep 3
# 
# pkill flameshot
```

</details>

And I got the following:

<details>
<summary>The output I got after running the <code>path_option.sh</code></summary>

> ```
> >> Nonexistent directory is not a valid path
> Invalid path, it must be a real path in the system. See flameshot --help.
> >> Saved relative to PWD
> Invalid path, it must be a real path in the system. See flameshot --help.
> >> Absolute paths work too
> Invalid path, it must be a real path in the system. See flameshot --help.
> >> Redundancy in the path will be removed
> Invalid path, it must be a real path in the system. See flameshot --help.
> >> If the destionation is a directory, the file name is generated from strf from the config
> >> If the output file has no suffix, it will be added
> Invalid path, it must be a real path in the system. See flameshot --help.
> >> If the suffix does not match the output format, it will be overwritten
> Invalid path, it must be a real path in the system. See flameshot --help.
> >> If the destination path exists, it will have _NUM appended to the base name
> Invalid path, it must be a real path in the system. See flameshot --help.
> >> We will do that again but without a suffix
> Invalid path, it must be a real path in the system. See flameshot --help.
> ```

</details>

These show that regardless of what happens, we always get the same exact error.

veracioux commented:
> I tried with the AppImage from this PR, and also commented out all lines of `path_option.sh` that were not necessary in my opinion:

Sure the lines that start the daemon and kill it are unnecessary and good on you for removing them. But I would still create an output directory so home doesn't get filled with junk. But I will change the destination to somewhere under /tmp.

I'm not sure why you are getting the invalid path errors, but did you try to run the script unmodified?

mmahmoudian commented:
There are two reason I removed those lines:
1. I was using the AppImage to test this, so the `flameshot &` wouldn't work, also the process name is not `flameshot` anymore and for that reason the `pkill flameshot` had to be changed to `pkill -i flameshot`
2. I didn't want to mess with the `${HOME}` variable and wanted to test with in a real environment (removing junk files is actually very easy considering their creating time, but not file was generated anyways)

I didn't try the unmodified. If you cannot figure out what the issue can be, I have to put things in a clean VM, compile your branch and then run the unmodified shell script. Let me know if you think this step is necessary for testing.

veracioux commented:
No idea why it doesn't work. I will try the appimage myself, and then report my findings. I will also modify the test script.

You don't need have to do anything for now. Thanks for your help.

veracioux commented:
@mmahmoudian When I try to download the appimage version, I keep getting a 404 error. Will try again later.

But I think I know what the problem is on your end. Both the daemon and the subcommands have to be run with the same version of flameshot. You probably ran the daemon from the appimage and the subcommands via the flameshot that was installed on your system. Please run the newest test script, following the comments in it. I emphasize that you should launch the daemon manually, using the proper version of flameshot (in your case, the appimage you downloaded). Don't rely on it being launched automatically, the versions might not match.

I didn't have this problem because I use a program to automatically manage my PATH variable. I should have disabled this and made it explicit that the daemon and subcommands need to use matching versions of flameshot. Also the test script was a bit rushed and unclear so I apologize for that.

veracioux commented:
@mmahmoudian I have managed to download the appimage. What I have observed is that everything works except when the specified path is relative. This is because the appimage extracts itself somewhere, so the directory of the appimage is not the same as the working directory of the extracted flameshot executable that gets run.

I need advice from someone who is more knowledgeable about appimages. The way I see it currently, there is no workaround that is intuitive to the user, i.e. appimage will have to be unsupported in this feature.

Still, I would like to ask you to run the test script and confirm that you are getting the same behavior.

veracioux commented:
@borgmanJeremy Thanks. Should be fixed now.

borgmanJeremy commented:
Okay great, anything remaining before merging?

veracioux commented:
@borgmanJeremy Yes, I still would like to do some more testing and add a few more things possibly. I was a little busy to do it lately, but it should be ready to merge in the upcoming days.

veracioux commented:
I made some small improvements:
- Default path is now changed only when the GUI dialog is used, the `--path` option won't change it
- Changed the help text for `--path` which partly addresses @mmahmoudian's [comment](https://github.com/flameshot-org/flameshot/pull/1815#issuecomment-895955655). The remainder I will address in another PR.
- The image format is now chosen based on the file extension
